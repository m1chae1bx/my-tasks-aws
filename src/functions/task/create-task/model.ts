import { JSONSchemaType, ajv } from "@libs/generic/util";

export interface CreateTaskRequest {
  pathParameters: {
    listId: string;
  };
  body: {
    name: string;
    desc?: string;
    dueDate?: string;
  };
}

export const createTaskRequestSchema: JSONSchemaType<CreateTaskRequest> = {
  $schema: "http://json-schema.org/draft-07/schema#",
  type: "object",
  properties: {
    pathParameters: {
      type: "object",
      properties: {
        listId: {
          type: "string",
        },
      },
      required: ["listId"],
    },
    body: {
      type: "object",
      properties: {
        name: {
          type: "string",
          maxLength: 200,
        },
        desc: {
          type: "string",
          nullable: true,
          maxLength: 400,
        },
        dueDate: {
          type: "string",
          nullable: true,
          format: "date-time",
        },
      },
      required: ["name"],
    },
  },
  required: ["pathParameters", "body"],
};

export const validateRequest = ajv.compile(createTaskRequestSchema);
