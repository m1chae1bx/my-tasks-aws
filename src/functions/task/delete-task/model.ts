import { JSONSchemaType, ajv } from "/opt/nodejs/util";

export interface DeleteTaskRequest {
  pathParameters: {
    listId: string;
    taskId: string;
  };
}

export const deleteTaskRequestSchema: JSONSchemaType<DeleteTaskRequest> = {
  $schema: "http://json-schema.org/draft-07/schema#",
  type: "object",
  properties: {
    pathParameters: {
      type: "object",
      properties: {
        listId: {
          type: "string",
        },
        taskId: {
          type: "string",
        },
      },
      required: ["listId", "taskId"],
    },
  },
  required: ["pathParameters"],
};

export const validateRequest = ajv.compile(deleteTaskRequestSchema);
