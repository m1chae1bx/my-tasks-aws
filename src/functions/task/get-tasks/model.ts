import { GetTasksQuery } from "/opt/nodejs/task.model";
import { JSONSchemaType, ajv } from "/opt/nodejs/util";

export interface GetTasksRequest {
  pathParameters: {
    listId: string;
  };
  queryStringParameters?: GetTasksQuery;
}

export const getTasksRequestSchema: JSONSchemaType<GetTasksRequest> = {
  $schema: "http://json-schema.org/draft-07/schema#",
  type: "object",
  properties: {
    pathParameters: {
      type: "object",
      properties: {
        listId: {
          type: "string",
        },
      },
      required: ["listId"],
    },
    queryStringParameters: {
      type: "object",
      properties: {
        name: {
          type: "string",
          nullable: true,
        },
        dueDate: {
          type: "string",
          enum: ["today", "tomorrow", "upcoming", "overdue", "unplanned"],
          nullable: true,
        },
        today: {
          type: "string",
          nullable: true,
          format: "date-time",
        },
        includeCompleted: {
          type: "string",
          enum: ["true"],
          nullable: true,
        },
      },
      nullable: true,
      if: { properties: { dueDate: { type: "string" } } },
      then: { required: ["today"] },
    },
  },
  required: ["pathParameters"],
};

export const validateRequest = ajv.compile(getTasksRequestSchema);
