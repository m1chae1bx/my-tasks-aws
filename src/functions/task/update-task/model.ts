import { JSONSchemaType, ajv } from "@libs/generic/util";

export interface UpdateTaskRequest {
  pathParameters: {
    listId: string;
    taskId: string;
  };
  body: {
    id: string;
    name: string;
    isCompleted?: boolean;
    desc?: string;
    dueDate?: string;
  };
}

export const updateTaskRequestSchema: JSONSchemaType<UpdateTaskRequest> = {
  $schema: "http://json-schema.org/draft-07/schema#",
  type: "object",
  properties: {
    pathParameters: {
      type: "object",
      properties: {
        listId: {
          type: "string",
        },
        taskId: {
          type: "string",
        },
      },
      required: ["listId", "taskId"],
    },
    body: {
      type: "object",
      properties: {
        id: {
          type: "string",
        },
        name: {
          type: "string",
          maxLength: 200,
        },
        isCompleted: {
          type: "boolean",
          nullable: true,
        },
        desc: {
          type: "string",
          nullable: true,
          maxLength: 400,
        },
        dueDate: {
          type: "string",
          nullable: true,
          format: "date-time",
        },
      },
      required: ["id", "name"],
    },
  },
  required: ["pathParameters", "body"],
  if: {
    properties: {
      pathParameters: {
        type: "object",
        properties: {
          taskId: {
            type: "string",
          },
        },
      },
    },
  },
  then: {
    properties: {
      body: {
        type: "object",
        properties: {
          id: {
            const: { $data: "/pathParameters/taskId" },
          },
        },
      },
    },
  },
};

export const validateRequest = ajv.compile(updateTaskRequestSchema);
