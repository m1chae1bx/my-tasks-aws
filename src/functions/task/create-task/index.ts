import { APIGatewayProxyEvent, APIGatewayProxyResult } from "aws-lambda";
import { validateRequest } from "./model";
import { Task } from "/opt/nodejs/task.model";
import { genericErrorHandler } from "/opt/nodejs/util";

export const handler = async (
  event: APIGatewayProxyEvent
): Promise<APIGatewayProxyResult> => {
  try {
    const request = {
      pathParameters: event.pathParameters,
      body: event.body ? JSON.parse(event.body) : {},
    };

    if (!validateRequest(request)) {
      return {
        statusCode: 400,
        body: JSON.stringify({
          message: "Bad request",
          errors: validateRequest.errors,
        }),
      };
    }

    const { name, dueDate, desc } = request.body;
    const { listId } = request.pathParameters;
    const task = new Task(
      listId,
      name,
      false,
      dueDate ? new Date(dueDate) : undefined,
      desc
    );

    const id = await task.save();
    const response = { message: `Task '${name}' was created successfully`, id };
    return {
      statusCode: 200,
      body: JSON.stringify(response),
    };
  } catch (error) {
    return genericErrorHandler(
      error,
      "An error occurred while creating the task. Please try again later."
    );
  }
};
