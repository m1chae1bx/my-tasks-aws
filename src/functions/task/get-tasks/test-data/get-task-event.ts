import { baseAPIGatewayProxyEvent } from "@test-utils/base-objects";

export const getTasksEvent = {
  ...baseAPIGatewayProxyEvent,
  pathParameters: {
    listId: "test-list-id",
  },
  queryStringParameters: {
    name: "test-name",
    dueDate: "2020-01-01",
    includeCompleted: "true",
  },
};
