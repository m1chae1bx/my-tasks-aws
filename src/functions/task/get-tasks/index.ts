import { APIGatewayProxyEvent, APIGatewayProxyResult } from "aws-lambda";
import { validateRequest } from "./model";
import { Task } from "/opt/nodejs/task.model";
import { genericErrorHandler } from "/opt/nodejs/util";

export const handler = async (
  event: APIGatewayProxyEvent
): Promise<APIGatewayProxyResult> => {
  try {
    const request = {
      pathParameters: event.pathParameters,
      queryStringParameters: event.queryStringParameters,
    };
    if (!validateRequest(request)) {
      return {
        statusCode: 400,
        body: JSON.stringify({
          message: "Bad request",
          errors: validateRequest.errors,
        }),
      };
    }

    const { listId } = request.pathParameters;
    const tasks = await Task.getAll(listId, request.queryStringParameters);
    return {
      statusCode: 200,
      body: JSON.stringify(tasks),
    };
  } catch (error) {
    return genericErrorHandler(
      error,
      "An error occurred while getting the tasks. Please try again later."
    );
  }
};
