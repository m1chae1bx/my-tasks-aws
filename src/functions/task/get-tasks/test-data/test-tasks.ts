import { Task } from "/opt/nodejs/task.model";

export const testTasks: Partial<Task>[] = [
  {
    id: "test-task-id-1",
    name: "test-task-name-2",
    desc: "test-task-desc",
    isCompleted: false,
    dueDate: new Date("2020-01-01T00:00:00.000Z"),
    listId: "test-list-id",
  },
  {
    id: "test-task-id-2",
    name: "test-task-name-2",
    desc: "test-task-desc",
    isCompleted: true,
    dueDate: new Date("2020-01-01T00:00:00.000Z"),
    listId: "test-list-id",
  },
];
