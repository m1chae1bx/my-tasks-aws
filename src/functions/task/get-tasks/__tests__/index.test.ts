import { ajvError } from "@test-utils/base-objects";
import { APIGatewayProxyResult } from "aws-lambda";
import { handler } from "..";
import { validateRequest } from "../model";
import { getTasksEvent } from "../test-data/get-task-event";
import { testTasks } from "../test-data/test-tasks";
import { Task } from "@libs/task";
import { genericErrorHandler } from "@libs/generic/util";

jest.mock("../model", () => ({
  validateRequest: jest.fn(),
}));

jest.mock("@libs/task", () => ({
  Task: {
    getAll: jest.fn(),
  },
}));

jest.mock("@libs/generic/util", () => ({
  genericErrorHandler: jest.fn(),
}));

describe("get-tasks", () => {
  describe("happy path", () => {
    let result: APIGatewayProxyResult;

    beforeAll(async () => {
      (validateRequest as unknown as jest.Mock).mockReturnValueOnce(true);
      (Task.getAll as jest.Mock).mockResolvedValueOnce(testTasks);
      result = await handler(getTasksEvent);
    });

    it("should return a 200 response", () => {
      expect(result.statusCode).toBe(200);
    });

    it("should return the tasks", () => {
      expect(result.body).toMatchSnapshot();
      testTasks.forEach((task) => expect(result.body).toContain(task.id));
    });
  });

  describe("sad path - invalid request", () => {
    let result: APIGatewayProxyResult;

    beforeAll(async () => {
      (validateRequest as unknown as jest.Mock).mockReturnValueOnce(false);
      validateRequest.errors = ajvError;
      result = await handler(getTasksEvent);
    });

    it("should return a 400 response", () => {
      expect(result.statusCode).toBe(400);
    });

    it("should return the error", () => {
      expect(result.body).toMatchSnapshot();
    });
  });

  // sad path - error while getting tasks
  describe("sad path - error while getting tasks", () => {
    let result: APIGatewayProxyResult;

    beforeAll(async () => {
      (validateRequest as unknown as jest.Mock).mockReturnValueOnce(true);
      (Task.getAll as jest.Mock).mockRejectedValueOnce(new Error("test error"));
      (genericErrorHandler as unknown as jest.Mock).mockReturnValueOnce({
        statusCode: 500,
        body: "test error",
      });
      result = await handler(getTasksEvent);
    });

    it("should return a 500 response", () => {
      expect(result.statusCode).toBe(500);
    });

    it("should return the error", () => {
      expect(result.body).toMatchSnapshot();
    });
  });
});
