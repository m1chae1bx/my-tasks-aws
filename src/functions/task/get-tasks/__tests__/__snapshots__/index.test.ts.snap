// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`get-tasks happy path should return the tasks 1`] = `"[{\\"id\\":\\"test-task-id-1\\",\\"name\\":\\"test-task-name-2\\",\\"desc\\":\\"test-task-desc\\",\\"isCompleted\\":false,\\"dueDate\\":\\"2020-01-01T00:00:00.000Z\\",\\"listId\\":\\"test-list-id\\"},{\\"id\\":\\"test-task-id-2\\",\\"name\\":\\"test-task-name-2\\",\\"desc\\":\\"test-task-desc\\",\\"isCompleted\\":true,\\"dueDate\\":\\"2020-01-01T00:00:00.000Z\\",\\"listId\\":\\"test-list-id\\"}]"`;

exports[`get-tasks sad path - error while getting tasks should return the error 1`] = `"test error"`;

exports[`get-tasks sad path - invalid request should return the error 1`] = `"{\\"message\\":\\"Bad request\\",\\"errors\\":[{\\"keyword\\":\\"\\",\\"instancePath\\":\\"\\",\\"schemaPath\\":\\"\\",\\"params\\":{},\\"message\\":\\"test error\\"}]}"`;
